### PR TITLE
Keepers: Support registry 1.2 sync

### PR DESCRIPTION
**Pull request by infiloop2**
Support full sync as well as logs of both registries

Previously our code use to interact with separate gethwrappers. Now i created a separate wrapper which wraps the two contract versions and rest of the code does not need to interact with individual gethwrappers

### CODE REVIEW

File: core/services/keeper/registry_synchronizer_sync.go [hunk 1]
```diff
@@ -58,20 +67,80 @@ func (rs *RegistrySynchronizer) addNewUpkeeps(reg Registry) error {
...
+			newUpkeeps = append(newUpkeeps, upkeepID)
+		}
+	}
+	rs.batchSyncUpkeepsOnRegistry(reg, newUpkeeps)
+
+	// All upkeeps in existingUpkeepIDs, not in activeUpkeepIDs should be deleted
+	activeSet := make(map[*big.Int]bool)
+	for _, upkeepID := range activeUpkeepIDs {
+		activeSet[upkeepID] = true
+	}
```
Comment (cmalec) [inline core/services/keeper/registry_synchronizer_sync.go:122]: here you are looping thru `activeUpkeepIDs` put there is already a loop for that at line 111. we could do this work at the same time i think.

File: core/services/keeper/registry_synchronizer_sync.go [hunk 2]
```diff
@@ -58,20 +67,80 @@ func (rs *RegistrySynchronizer) addNewUpkeeps(reg Registry) error {
...
+	canceled := make([]int64, 0)
+	for _, upkeepID := range existingUpkeepIDs {
```
Comment (cmalec) [inline core/services/keeper/registry_synchronizer_sync.go:124]: same idea with line 108

File: core/services/keeper/registry_interface.go [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+func getRegistryVersion(contract1_1 *registry1_1.KeeperRegistry) (*RegistryVersion, error) {
+	// Use registry 1_1 wrapper to get version information
+	typeAndVersion, err := contract1_1.TypeAndVersion(nil)
+	if err != nil {
+		jsonErr := evmclient.ExtractRPCError(err)
+		if jsonErr != nil {
+			// Version 1.0 does not support typeAndVersion interface, hence gives a json error on this call
+			version := RegistryVersion_1_0
+			return &version, nil
```
Comment (infiloop2) [inline core/services/keeper/registry_interface.go:70]: @RyanRHall's note: If someone provides a bad address (not even a registry), we should try to handle it gracefully

File: core/services/keeper/registry_interface.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+	// only applicable on 1.0 and 1.1
+	switch rw.Version {
+	case RegistryVersion_1_0, RegistryVersion_1_1:
+		return rw.contract1_1.GetCanceledUpkeepList(opts)
+	default:
+		return *new([]*big.Int), errors.Errorf("Registry version %d does not support GetCanceledUpkeepList", rw.Version)
+	}
+}
+
+func (rw *RegistryWrapper) GetActiveUpkeepIDs(opts *bind.CallOpts, startIndex *big.Int, maxCount *big.Int) ([]*big.Int, error) {
```
Comment (infiloop2) [inline core/services/keeper/registry_interface.go:110]: @RyanRHall 's note: These can be potentially merged with some abstraction layer

File: core/services/keeper/registry_interface_logs.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+				registry1_2.KeeperRegistryConfigSet{}.Topic():         nil,
+				registry1_2.KeeperRegistryUpkeepCanceled{}.Topic():    nil,
+				registry1_2.KeeperRegistryUpkeepRegistered{}.Topic():  nil,
+				registry1_2.KeeperRegistryUpkeepPerformed{}.Topic():   nil,
+				registry1_2.KeeperRegistryUpkeepGasLimitSet{}.Topic(): nil,
+			},
+			MinIncomingConfirmations: minIncomingConfirmations,
+		}, nil
+	default:
+		return nil, errors.Errorf("registry version not supported %d", rw.Version)
```
Comment (infiloop2) [inline core/services/keeper/registry_interface_logs.go:44]: @RyanRHall's note: Potentially make a helper function to get this generic error

**Review by infiloop2 — COMMENTED** [review]
Adding some comments for @RyanRHall

File: core/services/keeper/registry_interface_logs.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+		if !ok {
+			return nil, errors.Errorf("expected UpkeepRegistered log but got %T", broadcastedLog)
+		}
+		return broadcastedLog.Id, nil
+	default:
+		return nil, errors.Errorf("Registry version %d does not support GetUpkeepIdFromRegistrationLog", rw.Version)
+	}
+}
+
+func (rw *RegistryWrapper) GetUpkeepIdFromRawRegistrationLog(rawLog types.Log) (*big.Int, error) {
```
Comment (infiloop2) [inline core/services/keeper/registry_interface_logs.go:86]: this is only needed for test, might make sense to move to helper.test

File: core/services/keeper/registry_synchronizer_sync.go [hunk 7]
```diff
@@ -94,16 +160,16 @@ func (rs *RegistrySynchronizer) syncUpkeepWithCallback(registry Registry, upkeep
 	}
 }
 
-func (rs *RegistrySynchronizer) syncUpkeep(registry Registry, upkeepID int64) error {
-	upkeepConfig, err := rs.contract.GetUpkeep(nil, big.NewInt(upkeepID))
+func (rs *RegistrySynchronizer) syncUpkeep(registry Registry, upkeepID *big.Int) error {
+	upkeepConfig, err := rs.registryWrapper.GetUpkeep(nil, upkeepID)
```
Comment (infiloop2) [inline core/services/keeper/registry_synchronizer_sync.go:164]: @RyanRHall's comment: We should also sync the last keeper here

File: core/services/keeper/registry_synchronizer_core.go [hunk 8]
```diff
@@ -59,14 +58,15 @@ type RegistrySynchronizer struct {
...
 	mailRoom := MailRoom{
-		mbUpkeepCanceled:   utils.NewMailbox[log.Broadcast](50),
-		mbSyncRegistry:     utils.NewMailbox[log.Broadcast](1),
-		mbUpkeepPerformed:  utils.NewMailbox[log.Broadcast](300),
-		mbUpkeepRegistered: utils.NewMailbox[log.Broadcast](50),
+		mbUpkeepCanceled:    utils.NewMailbox[log.Broadcast](50),
+		mbSyncRegistry:      utils.NewMailbox[log.Broadcast](1),
+		mbUpkeepPerformed:   utils.NewMailbox[log.Broadcast](300),
+		mbUpkeepRegistered:  utils.NewMailbox[log.Broadcast](50),
+		mbUpkeepGasLimitSet: utils.NewMailbox[log.Broadcast](50),
```
Comment (infiloop2) [inline core/services/keeper/registry_synchronizer_core.go:65]: @RyanRHall 's comment: Let's increase these by a factor of 10 /100 right now.  During load testing let's try to understand the exact limits here

File: core/services/keeper/registry_synchronizer_process_logs.go [hunk 9]
```diff
@@ -5,17 +5,17 @@ import (
...
 
 func (rs *RegistrySynchronizer) processLogs() {
 	wg := sync.WaitGroup{}
-	wg.Add(4)
+	wg.Add(5)
 	go rs.handleSyncRegistryLog(wg.Done)
 	go rs.handleUpkeepCanceledLogs(wg.Done)
 	go rs.handleUpkeepRegisteredLogs(wg.Done)
 	go rs.handleUpkeepPerformedLogs(wg.Done)
+	go rs.handleUpkeepGasLimitSetLogs(wg.Done)
```
Comment (infiloop2) [inline core/services/keeper/registry_synchronizer_process_logs.go:18]: @RyanRHall 's comments: Potentially think about separating rs.handleSyncRegistryLog & rest logs from being in the same wait group. Probably needs a different ticket for investigation

File: core/services/keeper/registry_synchronizer_process_logs.go [hunk 10]
```diff
@@ -150,23 +154,67 @@ func (rs *RegistrySynchronizer) handleUpkeepPerformed(broadcast log.Broadcast) {
...
+	}
+
+	err = rs.syncUpkeep(registry, upkeepID)
+	if err != nil {
+		rs.logger.With("error", err).Error("failed to sync upkeep, log: %v", broadcast.String())
+		return
+	}
+	if err := rs.logBroadcaster.MarkConsumed(broadcast); err != nil {
+		rs.logger.With("error", err).Errorf("unable to mark KeeperRegistryUpkeepGasLimitSet log as consumed, log: %v", broadcast.String())
+	}
```
Comment (infiloop2) [inline core/services/keeper/registry_synchronizer_process_logs.go:219]: Potentially try to share code between functions using generics

File: core/services/keeper/registry_synchronizer_sync.go [hunk 11]
```diff
@@ -43,13 +40,24 @@ func (rs *RegistrySynchronizer) syncRegistry() (Registry, error) {
...
 }
 
-func (rs *RegistrySynchronizer) addNewUpkeeps(reg Registry) error {
+func (rs *RegistrySynchronizer) fullSyncUpkeeps(reg Registry) error {
+	switch rs.registryWrapper.Version {
+	case RegistryVersion_1_0, RegistryVersion_1_1:
+		return rs.fullSyncUpkeeps1_1(reg)
+	case RegistryVersion_1_2:
+		return rs.fullSyncUpkeeps1_2(reg)
+	}
```
Comment (infiloop2) [inline core/services/keeper/registry_synchronizer_sync.go:49]: try to combine these into the same abstraction

**Review by infiloop2 — COMMENTED** [review]
more comments

File: core/services/keeper/registry_interface.go (hunk 3 above)
Comment (infiloop2) [reply core/services/keeper/registry_interface.go:70]: Created story 38155 to handle this

File: core/services/keeper/registry_synchronizer_core.go (hunk 8 above)
Comment (infiloop2) [reply core/services/keeper/registry_synchronizer_core.go:65]: created 38160 to track this

File: core/services/keeper/registry_synchronizer_process_logs.go (hunk 9 above)
Comment (infiloop2) [reply core/services/keeper/registry_synchronizer_process_logs.go:18]: 38161 for investigation

File: core/services/keeper/registry_synchronizer_sync.go (hunk 7 above)
Comment (infiloop2) [reply core/services/keeper/registry_synchronizer_sync.go:113]: added

File: core/services/keeper/registry_interface.go [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+
+	return &RegistryWrapper{
+		Address:     address,
+		Version:     *version,
+		contract1_1: contract1_1,
+		contract1_2: contract1_2,
+	}, nil
+}
+
+func getRegistryVersion(contract1_1 *registry1_1.KeeperRegistry) (*RegistryVersion, error) {
```
Comment (RyanRHall) [inline core/services/keeper/registry_interface.go:62]: Shouldn't this be like...
```go
func getRegistryVersion(contract *type_and_version.TypeAndVersion) (*RegistryVersion, error) {
```

File: core/services/keeper/registry_interface.go [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+	}
+}
+
+func getUnsupportedVersionError(functionName string, version RegistryVersion) error {
+	return errors.Errorf("Registry version %d does not support %s", version, functionName)
+}
+
+func (rw *RegistryWrapper) GetActiveUpkeepIDs(opts *bind.CallOpts) ([]*big.Int, error) {
+	switch rw.Version {
+	case RegistryVersion_1_0, RegistryVersion_1_1:
```
Comment (RyanRHall) [inline core/services/keeper/registry_interface.go:92]: this looks like a wonderful abstraction to me :+1:

File: core/services/keeper/registry_synchronizer_core.go [hunk 14]
```diff
@@ -22,15 +20,16 @@ var (
...
+	mbSyncRegistry      *utils.Mailbox[log.Broadcast]
+	mbUpkeepPerformed   *utils.Mailbox[log.Broadcast]
+	mbUpkeepRegistered  *utils.Mailbox[log.Broadcast]
+	mbUpkeepGasLimitSet *utils.Mailbox[log.Broadcast]
 }
 
 type RegistrySynchronizerOptions struct {
 	Job                      job.Job
-	Contract                 *keeper_registry_wrapper1_1.KeeperRegistry
+	RegistryWrapper          *RegistryWrapper
```
Comment (RyanRHall) [inline core/services/keeper/registry_synchronizer_core.go:32]: just wondering if this _needs_ to be a pointer?

File: core/services/keeper/registry_synchronizer_process_logs.go [hunk 15]
```diff
@@ -112,12 +114,14 @@ func (rs *RegistrySynchronizer) HandleUpkeepRegistered(broadcast log.Broadcast,
 	if was {
 		return
 	}
-	broadcastedLog, ok := broadcast.DecodedLog().(*keeper_registry_wrapper1_1.KeeperRegistryUpkeepRegistered)
-	if !ok {
-		rs.logger.AssumptionViolationf("expected UpkeepRegistered log but got %T", broadcastedLog)
+
+	upkeepID, err := rs.registryWrapper.GetUpkeepIdFromRegistrationLog(broadcast)
```
Comment (RyanRHall) [inline core/services/keeper/registry_synchronizer_process_logs.go:118]: man I love how clean these changes are :tada:

File: core/services/keeper/registry_synchronizer_process_logs.go [hunk 16]
```diff
@@ -151,23 +155,67 @@ func (rs *RegistrySynchronizer) handleUpkeepPerformed(broadcast log.Broadcast) {
...
+	}
+	for {
+		broadcast, exists := rs.mailRoom.mbUpkeepGasLimitSet.Retrieve()
+		if !exists {
+			return
+		}
+		rs.handleUpkeepGasLimitSet(broadcast, registry)
+	}
+}
+func (rs *RegistrySynchronizer) handleUpkeepGasLimitSet(broadcast log.Broadcast, registry Registry) {
```
Comment (RyanRHall) [inline core/services/keeper/registry_synchronizer_process_logs.go:195]: newline

File: core/services/keeper/registry_synchronizer_sync.go [hunk 17]
```diff
@@ -20,12 +20,9 @@ func (rs *RegistrySynchronizer) fullSync() {
...
 	}
-	if err := rs.addNewUpkeeps(registry); err != nil {
-		rs.logger.With("error", err).Error("failed to add new upkeeps during fullSyncing registry")
-		return
-	}
-	if err := rs.deleteCanceledUpkeeps(); err != nil {
-		rs.logger.With("error", err).Error("failed to delete canceled upkeeps during fullSyncing registry")
+
+	if err := rs.fullSyncUpkeeps(registry); err != nil {
+		rs.logger.With("error", err).Error("failed to sync upkeeps during fullSyncing registry")
```
Comment (RyanRHall) [inline core/services/keeper/registry_synchronizer_sync.go:25]: same Q about With / wrap

**Review by RyanRHall — CHANGES_REQUESTED** [review]

File: core/services/keeper/registry_interface.go [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+			if _, found := cancelledSet[i]; !found {
+				activeUpkeeps = append(activeUpkeeps, big.NewInt(i))
+			}
+		}
+		return activeUpkeeps, nil
+	case RegistryVersion_1_2:
+		// TODO (sc-37024): Get active upkeep IDs from contract in batches
+		return rw.contract1_2.GetActiveUpkeepIDs(opts, big.NewInt(0), big.NewInt(0))
+	default:
+		return *new([]*big.Int), getUnsupportedVersionError("GetActiveUpkeepIDs", rw.Version)
```
[the remark this answers is not in the document]
Comment (infiloop2) [reply core/services/keeper/registry_interface.go:117]: yes, not sure why i went with that in the first place

File: core/services/keeper/registry_synchronizer_sync.go (hunk 17 above)
Comment (infiloop2) [reply core/services/keeper/registry_synchronizer_sync.go:25]: story 38536 to do this separately

File: core/services/keeper/registry_interface.go (hunk 12 above)
Comment (infiloop2) [reply core/services/keeper/registry_interface.go:62]: ah interesting suggestion, I'll try generating gethwrapper for typeAndVersion and see if this works

Comment (infiloop2) [reply core/services/keeper/registry_interface.go]: done

File: core/services/keeper/registry_interface.go [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+		return &version, nil
+	case "KeeperRegistry 1.2.0":
+		version := RegistryVersion_1_2
+		return &version, nil
+	default:
+		return nil, errors.Errorf("Registry type and version %s not supported", typeAndVersion)
+	}
+}
+
+func getUnsupportedVersionError(functionName string, version RegistryVersion) error {
```
[the remark this answers is not in the document]
Comment (infiloop2) [reply core/services/keeper/registry_interface.go]: done

File: core/services/keeper/registry_synchronizer_process_logs.go [hunk 20]
```diff
@@ -151,23 +155,67 @@ func (rs *RegistrySynchronizer) handleUpkeepPerformed(broadcast log.Broadcast) {
 		return
 	}
 
-	log, ok := broadcast.DecodedLog().(*keeper_registry_wrapper1_1.KeeperRegistryUpkeepPerformed)
-	if !ok {
-		rs.logger.AssumptionViolationf("expected UpkeepPerformed log but got %T", log)
+	log, err := rs.registryWrapper.ParseUpkeepPerformedLog(broadcast)
+	if err != nil {
+		rs.logger.With("error", err).Error("Unable to fetch upkeep ID from performed log")
```
[the remark this answers is not in the document]
Comment (RyanRHall) [reply core/services/keeper/registry_synchronizer_process_logs.go:160]: cool :+1: seems like a good ticket for a new joiner

File: core/services/keeper/registry_synchronizer_sync.go (hunk 17 above)
Comment (RyanRHall) [reply core/services/keeper/registry_synchronizer_sync.go:25]: :+1:

**Review by RyanRHall — APPROVED** [review]

**Pull request merged**